### PR TITLE
Graph theory measures script

### PR DESCRIPTION
**Pull request by frheault**
This is still WIP, I am trying to figure out the rich_club and if my usage of length weighted matrix and streamline count weighted matrix is correct.

However, this is pretty much the format it will have. This script computes the main, relevant graph theory measures (GTM) from connectivity matrices.

It computes a single subject at the time (because it requires two ordered input) but can append to a son file to enable a population analysis.

Example to follow, this is for early discussion with Maxime/Manon/Sami

https://github.com/aestrivex/bctpy/wiki

### CODE REVIEW

**Comment by frheault** [comment]
@mdesco @arnaudbore This is ready for review.
I am still not familiar with the interpretation/analysis that will follow, but it is required by Sami's project.

I tried it with some data, but I think before merging someone (please not me, I hate Matlab, does Manon has it installed on her computer ?) should test it against the original https://sites.google.com/site/bctnet/Home/functions/BCT.zip?attredirects=0

Here is test data
[GTM_PR.zip](https://github.com/scilus/scilpy/files/4438706/GTM_PR.zip)

--------------------------------------
Here is a quick recap of the measure I chose and what they take as a input (not real code, I removed a lot of stuff just to summarize)
https://github.com/aestrivex/bctpy/wiki
```
centrality = bct.betweenness_wei(sc_matrix)
modularity = bct.modularity_finetune_und(sc_matrix)
assortativity = bct.assortativity_wei(sc_matrix)
participation = bct.participation_coef_sign(sc_matrix)
clustering = bct.clustering_coef_wu(sc_matrix)
degree = bct.degrees_und(sc_matrix)
module_degree_zscore = bct.module_degree_zscore(sc_matrix)
nodal_strength = bct.strengths_und(sc_matrix)
local_efficiency = bct.efficiency_wei(sc_matrix, local=True)
global_efficiency = bct.efficiency_wei(sc_matrix)
density = bct.density_und(sc_matrix)
rich_club = bct.rich_club_wu(sc_matrix)
path_length, edge_count = bct.distance_wei(len_matrix)
```

**Comment by arnaudbore** [comment]
Please add a test.

File: scripts/scil_compute_connectivity.py [hunk 1]
```diff
@@ -245,9 +244,11 @@ def main():
         for in_name, out_name in args.metrics:
             # Verify that all metrics are compatible with each other
             if not is_header_compatible(args.metrics[0][0], in_name):
-                continue
+                raise IOError('Metrics do not share a compatible header'.format(
```
Comment (arnaudbore) [inline scripts/scil_compute_connectivity.py:247]: forgot {} {} ?

File: scripts/scil_evaluate_connectivity_measures.py [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+        1059-1069.
+"""
+
+
+def _build_arg_parser():
+    p = argparse.ArgumentParser(description=__doc__,
+                                formatter_class=argparse.RawTextHelpFormatter,
+                                epilog=EPILOG)
+    p.add_argument('in_length_matrix',
+                   help='Input length weighted matrix.')
```
Comment (arnaudbore) [inline scripts/scil_evaluate_connectivity_measures.py:58]: add `(npy).` ? I know once you get there it should be obvious but I think it's still useful. Tell me what do you think.

**Comment by mdesco** [comment]
I cannot test anything at the moment because I cannot run the pip install -r requirements. 

I will ask Manon to test this.

**Comment by arnaudbore** [comment]
> I cannot test anything at the moment because I cannot run the pip install -r requirements.
> 
> I will ask Manon to test this.

**Comment by arnaudbore** [comment]
> I cannot test anything at the moment because I cannot run the pip install -r requirements.
Known issue [Here](https://github.com/scilus/scilpy/issues/198)

> 
> I will ask Manon to test this.

**Comment by GuillaumeTh** [comment]
I checked the code an lgtm.

**Comment by frheault** [comment]
@arnaudbore @mdesco While not perfect match, these measures do matches with Manon output of BCT. And to some extend I think they make sense with my preliminary interpretation (when I compare set vs. prob or when using a filtering mask)

Last round of review ?

File: scripts/scil_evaluate_connectivity_measures.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+        assert_outputs_exist(parser, args, args.out_json)
+    else:
+        logging.debug('Using --append_json, make sure to delete {} '
+                      'before re-launching a group analysis.'.format(
+                          args.out_json))
+
+    if args.append_json and args.overwrite:
+        parser.error('Cannot use the append option at the same time as '
+                     'overwrite.\nAmbiguous behavior, consider deleting the '
+                     'output json file first instead')
```
Comment (arnaudbore) [inline scripts/scil_evaluate_connectivity_measures.py:172]: .

File: scripts/scil_evaluate_connectivity_measures.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+    A network is commonly classified as small-world if sigma > 1.
+
+    Parameters
+    ----------
+    matrix : numpy.ndarray
+        A weighted undirected graph.
+    Returns
+    -------
+    smallworld : tuple of float
+        The small-work coefficients (omega & sigma)
```
Comment (arnaudbore) [inline scripts/scil_evaluate_connectivity_measures.py:102]: .

File: scripts/scil_evaluate_connectivity_measures.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+    add_json_args(p)
+    add_verbose_arg(p)
+    add_overwrite_arg(p)
+
+    return p
+
+
+def omega_sigma(matrix):
+    """Returns the small-world coefficients (omega & sigma) of a graph.
+    Omega ranges between -1 and 1. Values close to 0 mean the metrix
```
Comment (arnaudbore) [inline scripts/scil_evaluate_connectivity_measures.py:88]: metrix -> metrics, matrix ??

File: scripts/scil_compute_connectivity.py [hunk 6]
```diff
@@ -143,11 +143,10 @@ def _processing_wrapper(args):
         elif os.path.isfile(measure):
             metric_filename = measure
             if not is_header_compatible(metric_filename, sft):
-                logging.error('{} and {} do not have a compatible header'.format(
+                raise IOError('{} and {} do not have a compatible header'.format(
                     in_filename, metric_filename))
-                raise IOError
 
-            metric_data = nib.load(metric_filename).get_data()
+            metric_data = nib.load(metric_filename).get_fdata(dtype=np.float32)
```
Comment (arnaudbore) [inline scripts/scil_compute_connectivity.py:149]: np.float64 or default

**Pull request merged**